src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
 makeSem_ :: Name -> Q [Dec]
 makeSem_ = genFreer False
 
+------------------------------------------------------------------------------
 -- | Generates declarations and possibly signatures for functions to lift GADT
 -- constructors into 'Sem' actions.
 genFreer :: Bool -> Name -> Q [Dec]
-genFreer makeSigs tcName = do
-  -- The signatures for the generated definitions require FlexibleContexts.
-  isExtEnabled FlexibleContexts
-    >>= flip unless (fail "makeSem requires FlexibleContexts to be enabled")
-  hasTyFams <- isExtEnabled TypeFamilies
-
-  reify tcName >>= \case
-    TyConI (DataD _ _ _ _ cons _) -> do
-      sigs <- filter (const makeSigs) <$> mapM genSig cons
-      decs <- join <$> mapM genDecl cons
-      loc <- location
-
-      return $
-        [ TySynInstD ''DefiningModule
-            . TySynEqn [ConT tcName]
-            . LitT
-            . StrTyLit
-            $ loc_module loc
-        | hasTyFams
-        ] ++ sigs ++ decs
-
-    _ -> fail "makeSem expects a type constructor"
-
--- | Given the name of a GADT constructor, return the name of the corresponding
--- lifted function.
-getDeclName :: Name -> Name
-getDeclName = mkName . overFirst toLower . nameBase
- where
-  overFirst f (a : as) = f a : as
-  overFirst _ as       = as
+genFreer shouldMkSigs typeName = do
+  dtInfo <- reifyDatatype typeName
+  -- TODO: data families seem to work just fine - should they be restricted?
+  clInfos  <- traverse (mkCLInfo dtInfo) $ datatypeCons dtInfo
+  defModFI <- TySynInstD ''DefiningModule
+          .   TySynEqn [ConT $ datatypeName dtInfo]
+          .   LitT
+          .   StrTyLit
+          .   loc_module
+          <$> location
+  sigs     <- if shouldMkSigs then traverse genSig clInfos
+                              else return []
+  decs     <- traverse genDec clInfos
+
+  return $ defModFI : join (sigs:decs)
 
+-------------------------------------------------------------------------------
+-- | Info about constructor being lifted; use 'mkCLInfo' to create one.
+data ConLiftInfo = CLInfo {
+    typeName :: Name
+  , funName  :: Name
+  , conName  :: Name
+  , effTArgs :: [Type]
+  , mName    :: Name
+  , resArg   :: Type
+  , conCxt   :: Cxt
+  , conArgs  :: [Type]
+  } deriving Show
+
+------------------------------------------------------------------------------
+-- | Creates info about constructor being lifted from parent type's and it's
+-- own info.
+mkCLInfo :: DatatypeInfo -> ConstructorInfo -> Q ConLiftInfo
+-- mkCLInfo dtInfo conInfo = do
+mkCLInfo DatatypeInfo {
```

Comment:
I find this hard to read; maybe parens around the record unpacking would help?

```haskell
mkCLInfo
    (DatatypeInfo
      { datatypeName = typeName
      , ...
      })
    (ConstructorInfo
      { ...
      }) = do
  let funName = ...
```

Reply:
Ok